README.md
# LeetCode Time Tracker Extension

![icon](/dist/icon.png)



[![Build Status](https://travis-ci.com/VarthanV/LeetCode-Time-Tracker.svg?token=sPWc2kHqb9ZisQqfsF7y&branch=master)](https://travis-ci.com/VarthanV/LeetCode-Time-Tracker)

## Description

LeetCode TimeTracker Extension helps you to track the time that you take to solve a problem and helps you to make progress

## Installation

Install the extension from the Chrome WebStore <https://chrome.google.com/webstore/detail/leetcode-time-tracker/obcihoolahbncgakepoceagnjnfgghhl?hl=en&authuser=0>

## Users installing from Github

- Clone the Repo
- Select **Load Unpacked** Extension from the Extensions Page(Make sure Developers Options is enabled).
- Select the **dist** folder in the repo

You are good to go !

## Screenshots

 A timer badge that helps you to  keep an eye on time while solving problem.

![shot1](/screenshots/shot1.png)

See your statistics and list of problems you solved.

![shot4](/screenshots/shot4.png)

Sort by difficulty

![shot3](/screenshots/shot3.png)

Search Specific Problems

![shot5](/screenshots/shot5.png)

Export your progress as CSV for further use

![shot2](/screenshots/shot2.png)

## Features yet to come

- Dark mode
- Setting time limit yourselves
- Sort by Date (Completed)
- Supporting Platforms like HackerRank ,CodeChef

## Updated UI Teaser

![shot6](/screenshots/shot6.png)

## Dark Mode

 ![shot7](/screenshots/shot7.png)

The project is still in Developement but the Chrome Store extension has a stable build . For detailed changes see [CHANELOG.md](https://github.com/VarthanV/LeetCode-Time-Tracker/blob/master/CHANGELOG.md)

Incase you have worked on above features submit a Pull Request   [Pull Request](https://github.com/VarthanV/LeetCode-Time-Tracker/pulls)

If any issues ,Please submit one [New Issuse](https://github.com/VarthanV/LeetCode-Time-Tracker/issues)


Icons made by <a href="https://www.flaticon.com/authors/freepik" title="Freepik">Freepik</a> from <a href="https://www.flaticon.com/" title="Flaticon"> www.flaticon.com</a>


CHANGELOG.md
# 1.0.0 : June 27,2020

- Made the Initial Release

## 1.0.1 : June 28,2020 (Beta)

- Massive UI Change (Done by [KrishnaMoorthy12](https://github.com/KrishnaMoorthy12))

- Improved Filter Functionality
- Added the Functionality of Filtering by date
- Dark Mode
- Improved accessibility in Search Bar


src/__tests__/test.ts

import {getTimeasString} from '../helpers';
import { normalize } from '../helpers';


test('Returns 2m ' ,() =>{
    expect(getTimeasString(0,2,0)).toBe('2 minutes')
})
test('Returns 3 '  , () =>{
    
    expect(normalize(0o3)).toBe(3)
})

test('Returns 1hr 25min ' , () =>{
    expect(getTimeasString(0,25,1)).toBe('1hr 25minutes')
})





src/options.ts
 
import { Chart } from "chart.js";

function renderChart() {
    alert("hi")
    let easyCount = 0;
    let mediumCount = 0;
    let hardCount = 0;
  
    const items = localStorage.getItem("leetCodeExtensionDetails");
    if (items) {
      const parsedItem = JSON.parse(items);
      easyCount = parsedItem.easy.filter(
        (item) => typeof item.duplicateIndex == "number"
      ).length;
      mediumCount = parsedItem.medium.filter(
        (item) => typeof item.duplicateIndex == "number"
      ).length;
      hardCount = parsedItem.hard.filter(
        (item) => typeof item.duplicateIndex == "number"
      ).length;
    }
    //@ts-ignore
    var ctx = document.getElementById("myChart").getContext("2d");
    var myChart = new Chart(ctx, {
      type: "doughnut",
      data: {
        labels: ["Easy", "Medium", "Hard"],
        datasets: [
          {
            label: "# of Votes",
            data: [easyCount, mediumCount, hardCount],
            backgroundColor: [
              "rgb(68,160,72)",
              "rgb(239,113,9)",
              "rgb(233,32,99)",
            ],
          },
        ],
      },
      options: {},
    });
  }

document.addEventListener('DOMContentLoaded',renderChart);

src/background.ts
import { Problem, Action } from "./types";
import { getTimeasString } from "./helpers";

// Overriding Custom Window Object Since cannot am unable to extend the default window obj
let windowObj = <any>window;
let timeString;
let currentUrl;
let milisec = 0;
let sec = 0; /* holds incrementing value */
let min = 0;
let hour = 0;
let miliSecOut = 0;
let secOut = 0;
let minOut = 0;
let hourOut = 0;

// Setting Popup dynamically
chrome.browserAction.setBadgeBackgroundColor({ color: "#5CAD62" });


chrome.tabs.onActivated.addListener(function (activeInfo) {
  const activeTabId = activeInfo.tabId;
  chrome.tabs.get(activeTabId, function (tab) {
    currentUrl = tab.url;

  });
});

// Persisting Data Stuffst

const problem: Problem = {};
chrome.runtime.onMessage.addListener(function (
  request: Action,
  sender,
  sendResponse
) {
  if (request) {
    if (request.action == "setProblem") {
      problem.problemName = request.payload.problemName;
      problem.difficulty = request.payload.difficulty;
    } else if (request.action == "getProblem") {
      sendResponse(problem);
    } else if (request.action == "setTimer") {
      startStop();
    } else if (request.action == "getTimer") {
      sendResponse({ startstop: startstop });
    }
    else if (request.action === 'getCurrentTime') {

      sendResponse({
        miliSecOut: miliSecOut,
        hourOut: hourOut,
        secOut: secOut,
        minOut: minOut
      })
    }
  }
});

// Updating Timer Stuffs

let startstop = 0;
let x;
/* Toggle StartStop */
function startStop() {
  startstop = startstop + 1;
  let document = chrome.extension.getViews({ type: "popup" })[0].document;

  if (startstop == 1) {
    startTimer();
    document.getElementById("start").innerHTML = `<span class="material-icons material-icons-outlined">pause</span>
        <span class="btn-txt">Pause</span>`;
  } else if (startstop == 2) {
    document.getElementById("start").innerHTML = `<span class="material-icons material-icons-outlined">arrow_forward_ios </span>
        <span class="btn-txt">Start</span>`;
    startstop = 0;
    stopTimer();
  }
}

// Creating my Custom Window object since Typescript doesnt allow much flexiblity
windowObj.startStop = startStop;
windowObj.resetFunc = reset;
windowObj.saveData = saveData;

//Starts the Timer

function startTimer() {
  x = setInterval(timer, 10);
}

// Stops the timer
function stopTimer() {
  clearInterval(x);
}

// Declaring Variables

// Driver for Timer
function timer() {
  let popup = chrome.extension.getViews({ type: "popup" })[0];
  let document = popup && popup.document;
  miliSecOut = checkTime(milisec);
  secOut = checkTime(sec);
  minOut = checkTime(min);
  hourOut = checkTime(hour);

  milisec = ++milisec;

  if (milisec === 100) {
    milisec = 0;
    sec = ++sec;
  }

  if (sec == 60) {
    min = ++min;
    sec = 0;
  }

  if (min == 60) {
    min = 0;
    hour = ++hour;
  }

  // Updates DOM

  /*
  When the Popup is closed background does'nt have reference to the Popup's
  document object  we no need to render when the popup is not opened


  */

  if (document) {
    document.getElementById("milisec").innerHTML = miliSecOut.toString();
    document.getElementById("sec").innerHTML = secOut.toString();
    document.getElementById("min").innerHTML = minOut.toString();
    document.getElementById("hour").innerHTML = hourOut.toString();
  }

  // Update the badge Text
  timeString = getTimeasString(secOut, minOut, hourOut);

  chrome.browserAction.setBadgeText({ text: timeString });
}

// Checks time to add 0 or not
function checkTime(i) {
  if (i < 10) {
    i = "0" + i;
  }
  return i;
}

// Resets the timer
function reset() {
  stopTimer();
  chrome.browserAction.setBadgeText({ text: "" });

  let document = chrome.extension.getViews({ type: "popup" })[0].document;
  /*Reset*/
  startstop = 0;
  document.getElementById("start").innerHTML = `<span class="material-icons material-icons-outlined">arrow_forward_ios </span>
        <span class="btn-txt">Start</span>`;
  milisec = 0;
  sec = 0;
  min = 0;
  hour = 0;
  miliSecOut = 0;
  hourOut = 0;
  minOut = 0;
  secOut = 0;
  document.getElementById("milisec").innerHTML = "00";
  document.getElementById("sec").innerHTML = "00";
  document.getElementById("min").innerHTML = "00";
  document.getElementById("hour").innerHTML = "00";
}

function setData() {
  let data = JSON.parse(localStorage.getItem("leetCodeExtensionDetails"));
  let today = new Date();
  //@ts-ignore
  let dd = String(today.getDate()).padStart(2, "0");
  //@ts-ignore
  let mm = String(today.getMonth() + 1).padStart(2, "0"); //January is 0!
  let yyyy = today.getFullYear();

  const todayString = dd + "/" + mm + "/" + yyyy;
  const problemName = problem.problemName.split(".")[1];
  const dataMap: any = {
    problemName: problemName,
    difficulty: problem.difficulty,
    timeTaken: timeString,
    date: todayString,
    problemUrl: currentUrl,
  };
  let currentDiffProblemArray: object[] =
    data[problem.difficulty.toLowerCase()];

  let exists = currentDiffProblemArray.filter((item: any) => {
    return item.problemName === problemName;
  });
  let problemObj: any = exists[0];

  if (exists.length > 0) {
    // dataMap.problemName +=  problemObj.duplicateIndex.toString();
    problemObj.duplicateIndex += 1;
    const itemToFind = (item) => item.problemName == dataMap.problemName;
    let idx = currentDiffProblemArray.findIndex(itemToFind);
    data[problem.difficulty.toLowerCase()][idx] = problemObj;

    dataMap.problemName += ` (${problemObj.duplicateIndex})`;

    data[problemObj.difficulty.toLowerCase()].push(dataMap);
    localStorage.setItem("leetCodeExtensionDetails", JSON.stringify(data));
  } else {
    dataMap.duplicateIndex = 0;
    data[dataMap.difficulty.toLowerCase()].push(dataMap);
    let dataToSet = JSON.stringify(data);

    localStorage.setItem("leetCodeExtensionDetails", dataToSet);
  }
  alert("Saved!");
  chrome.runtime.sendMessage({ showGraph: true });
}

async function saveData() {
  await reset();
  if (!localStorage.getItem("leetCodeExtensionDetails")) {
    setInitialData();
    setData();
  } else {
    setData();
  }
}

function setInitialData() {
  var problemDetails = JSON.stringify({
    easy: [],
    medium: [],
    hard: [],
  });
  localStorage.setItem("leetCodeExtensionDetails", problemDetails);
}


src/content_script.ts
import { Action } from "./types";

const getDifficulty = () => {
  let diffDiv: HTMLDivElement = document.querySelector("div[diff]");
  if (diffDiv == null) {
    setTimeout(getDifficulty, 3000);
  } else {
    let difficulty = diffDiv.innerText;
    return difficulty;
  }
};

const getProblemName = () => {
  let problemTitleDiv: HTMLDivElement = document.querySelector(
    'div[data-cy="question-title"]'
  );

  if (problemTitleDiv == null) {
    setTimeout(getProblemName, 3000);
  } else {
    const problemName = problemTitleDiv.innerText;
    return problemName;
  }
};

const getDetails = () => {
  let problemName = getProblemName();
  let difficulty = getDifficulty();
  if (problemName && difficulty) {
    const action: Action = {
      action: "setProblem",
      payload: {
        //@ts-ignore
        difficulty: difficulty,
        problemName: problemName,
      },
    };
    //Send Extracted Data
    chrome.runtime.sendMessage(action);
  } else {
    setTimeout(getDetails, 3000);
  }
};
//Get Details
chrome.runtime.onMessage.addListener(function (request, sender, response) {
  getDetails();
});

//Driver Code

getDetails();


src/popup.ts
import { Problem, Action } from "./types";
import { Chart } from "chart.js";

//Dev Stuffs

// @ts-ignore
const console = chrome.extension.getBackgroundPage().console;
/*
Cache Dom

*/
const problemTitleDiv = document.querySelector(
  "#problem-title"
) as HTMLSpanElement;
const difficultyDiv = document.querySelector(
  "#problem-difficulty"
) as HTMLSpanElement;
const tableDiv = document.querySelector("#problem-table");
const difficultySelectorDiv = document.querySelector(
  "#difficulty-selector"
) as HTMLSelectElement;
const millisecondsDiv = document.getElementById("milisec") as HTMLSpanElement;
const secondsDiv = document.getElementById("sec") as HTMLSpanElement;
const minutesDiv = document.getElementById("min") as HTMLSpanElement;
const hoursDiv = document.getElementById("hour") as HTMLSpanElement;
const searchQueryDiv = document.getElementById(
  "search-text"
) as HTMLInputElement;

const datePickerDiv = document.querySelector(
  "#problem-date"
) as HTMLInputElement;

// Buttons Div

const startBtn = document.getElementById("start") as HTMLButtonElement;
const resetBtn = document.getElementById("reset") as HTMLButtonElement;
const saveBtn = document.getElementById("save") as HTMLButtonElement;
const exportToCSVButton = document.getElementById(
  "export-to-csv"
) as HTMLButtonElement;
const searchBtn = document.getElementById("search-btn") as HTMLButtonElement;
const clearBtn = document.getElementById("clear-btn") as HTMLButtonElement;
// Get Title of the Current Page and URL
const backgroundPage = chrome.extension.getBackgroundPage();
// Variable Declarations

let easyProblems;
let mediumProblems;
let hardProblems;
let selectedValue = "all";
let selectedDate = null;
let dateString = "";
//  Background Page

/*
Adding Event Listeners

*/
//@ts-ignore
startBtn.addEventListener("click", function () {
  //@ts-ignore
  backgroundPage.startStop();
});

resetBtn.addEventListener("click", function () {
  //@ts-ignore
  backgroundPage.resetFunc();
});

saveBtn.addEventListener("click", function () {
  //@ts-ignore
  backgroundPage.saveData();
});

searchBtn.addEventListener("click", function () {
  search();
});
exportToCSVButton.addEventListener("click", function () {
  if (selectedValue === "all") {
    export_table_to_csv(document, "leetcodestats.csv");
  } else if (selectedValue === "easy") {
    export_table_to_csv(document, "leetcode_easystats.csv");
  } else if (selectedValue === "medium") {
    export_table_to_csv(document, "leetcode_mediumstats.csv");
  } else if (selectedValue === "hard") {
    export_table_to_csv(document, "leetcode_hardstats.csv");
  }
});

difficultySelectorDiv.addEventListener("change", function () {
  selectedValue =
    difficultySelectorDiv.options[difficultySelectorDiv.selectedIndex].value;
  clearUI();
  if (selectedDate === null) {
    if (selectedValue == "all") {
      renderTable();
    } else if (selectedValue === "easy") {
      renderItem(easyProblems, selectedValue);
    } else if (selectedValue === "medium") {
      renderItem(mediumProblems, selectedValue);
    } else if (selectedValue == "hard") {
      renderItem(hardProblems, selectedValue);
    }
  }
  else {
    renderWithRespectToDate();
  }
});
datePickerDiv.addEventListener("change", function () {
  selectedDate = datePickerDiv.value;
  selectedDate = new Date(selectedDate);
  let dd = String(selectedDate.getDate()).padStart(2, "0");
  //@ts-ignore
  let mm = String(selectedDate.getMonth() + 1).padStart(2, "0"); //January is 0!
  let yyyy = selectedDate.getFullYear();

  dateString = dd + "/" + mm + "/" + yyyy;
  renderWithRespectToDate();
});

function renderWithRespectToDate() {
  let easyProblemsRender = easyProblems.filter((item) => {
    return item.date === dateString;
  });
  let mediumProblemsRender = mediumProblems.filter((item) => {
    return item.date === dateString;
  });

  let hardProblemsRender = hardProblems.filter((item) => {
    return item.date === dateString;
  });
  clearUI();
  if (selectedValue === "all") {
    renderItem(easyProblemsRender, "easy");
    renderItem(mediumProblemsRender, "medium");
    renderItem(hardProblemsRender, "hard");
  } else if (selectedValue === "easy") {
    renderItem(easyProblemsRender, "easy");
  } else if (selectedValue === "medium") {
    renderItem(mediumProblemsRender, "medium");
  } else if (selectedValue === "hard") {
    renderItem(hardProblemsRender, "hard");
  }
}
// Search Triggers when enter key is pressed
searchQueryDiv.addEventListener("keypress", function (e) {
  if (e.key === "Enter") {
    if (searchQueryDiv.value) {
      search();
    }
  }
});
clearBtn.addEventListener("click", function () {
  selectedDate = null;
  datePickerDiv.value = null;
  selectedValue = "all";
  difficultySelectorDiv.value = "all";
  clearUI();
  renderTable();
});

/*
Rendering DOM

*/

document.addEventListener("DOMContentLoaded", function () {
  renderTimerPage();
  renderChart();
  renderTable();
});

// Renders Timer
function renderTimerPage() {
  let problemDict: Problem = {};

  // Get the Problem's Name and update DOMM
  const action: Action = {
    action: "getProblem",
  };
  chrome.runtime.sendMessage(action, function (response: Problem) {
    problemDict.problemName = response.problemName;
    problemDict.difficulty = response.difficulty;
    console.log(problemDict);

    if (problemDict.problemName) {
      let name = problemDict.problemName.split(".")[1];
      let difficulty = problemDict.difficulty;
      problemTitleDiv.innerText = name;
      difficultyDiv.innerText = difficulty;

      difficultyDiv.classList.add(difficulty.toLowerCase());
    }
  });

  // Get the Current State and update DOM
  const timerAction: Action = {
    action: "getTimer",
  };
  chrome.runtime.sendMessage(timerAction, function (response) {
    if (response.startstop == 1) {
      document.getElementById(
        "start"
      ).innerHTML = `<span class="material-icons material-icons-outlined">pause</span>
        <span class="btn-txt">Pause</span>`;
    } else if (response.startstop == 2) {
      document.getElementById(
        "start"
      ).innerHTML = `<span class="material-icons material-icons-outlined">arrow_forward_ios </span>
        <span class="btn-txt">Start</span>`;
    }
  });

  //  Get how much time has elapsed
  const getCurrentTimeAction: Action = {
    action: "getCurrentTime",
  };
  chrome.runtime.sendMessage(getCurrentTimeAction, function (response) {
    const result = specialCase(
      response.miliSecOut,
      response.secOut,
      response.minOut,
      response.hourOut
    );
    if (result) {
      millisecondsDiv.innerText = "00";
      secondsDiv.innerText = "00";
      minutesDiv.innerText = "00";
      hoursDiv.innerText = "00";
    } else {
      millisecondsDiv.innerText = response.miliSecOut.toString();
      secondsDiv.innerText = response.secOut.toString();
      minutesDiv.innerText = response.minOut.toString();
      hoursDiv.innerText = response.hourOut.toString();
    }
  });
}

function specialCase(
  milisec: number,
  second: number,
  minute: number,
  hour: number
) {
  if (milisec === 0 && second === 0 && minute === 0 && hour === 0) {
    return true;
  }
  return false;
}
// Listeners

// Listens for Changes and update UI  Accordingly
chrome.runtime.onMessage.addListener(function (request) {
  if (request.showGraph) {
    renderChart();
    clearUI();
    renderTable();
  }
});

/*
Clear's UI
*/

function clearUI() {
  tableDiv.innerHTML = null;
}

/*
Renders Chart

*/
//@ts-ignore

function renderChart() {
  let easyCount = 0;
  let mediumCount = 0;
  let hardCount = 0;

  const items = localStorage.getItem("leetCodeExtensionDetails");
  if (items) {
    const parsedItem = JSON.parse(items);
    easyCount = parsedItem.easy.filter(
      (item) => typeof item.duplicateIndex == "number"
    ).length;
    mediumCount = parsedItem.medium.filter(
      (item) => typeof item.duplicateIndex == "number"
    ).length;
    hardCount = parsedItem.hard.filter(
      (item) => typeof item.duplicateIndex == "number"
    ).length;
  }
  //@ts-ignore
  var ctx = document.getElementById("myChart").getContext("2d");
  var myChart = new Chart(ctx, {
    type: "doughnut",
    data: {
      labels: ["Easy", "Medium", "Hard"],
      datasets: [
        {
          label: "# of Votes",
          data: [easyCount, mediumCount, hardCount],
          backgroundColor: [
            "rgb(68,160,72)",
            "rgb(239,113,9)",
            "rgb(233,32,99)",
          ],
        },
      ],
    },
    options: {},
  });
}

/*
Renders Table
*/
function renderTable() {
  const items = localStorage.getItem("leetCodeExtensionDetails");
  if (items) {
    const parsed = JSON.parse(items);
    easyProblems = parsed.easy;
    mediumProblems = parsed.medium;
    hardProblems = parsed.hard;
    renderItem(easyProblems, "easy");
    renderItem(mediumProblems, "medium");
    renderItem(hardProblems, "hard");
  }
}

/*
Renders Each Individual Table item

*/
function renderItem(items, classname?) {
  for (const item of items) {
    const row = document.createElement("tr");
    const timeTaken = document.createElement("td");
    const problemLink = document.createElement("a");
    const problemName = document.createElement("td");

    const difficulty = document.createElement("td");
    const dateSolved = document.createElement("td");
    problemLink.innerText = item.problemName;
    problemLink.href = item.problemUrl;
    problemLink.setAttribute("target", "blank");
    problemName.appendChild(problemLink);
    difficulty.innerText = item.difficulty;
    difficulty.classList.add(
      classname !== undefined ? classname : item.difficulty.toLowerCase()
    );

    timeTaken.innerText = item.timeTaken;
    dateSolved.innerText = item.date;

    // Appending to the Row
    row.appendChild(problemName);
    row.appendChild(difficulty);
    row.appendChild(timeTaken);
    row.appendChild(dateSolved);

    tableDiv.appendChild(row);
  }
}

/*
CSV Helpers
Credits : https://jsfiddle.net/gengns/j1jm2tjx/

*/
function download_csv(csv, filename) {
  let csvFile;
  let downloadLink;

  // CSV FILE
  csvFile = new Blob([csv], { type: "text/csv" });

  // Download link
  downloadLink = document.createElement("a");

  // File name
  downloadLink.download = filename;

  // We have to create a link to the file
  downloadLink.href = window.URL.createObjectURL(csvFile);

  // Make sure that the link is not displayed
  downloadLink.style.display = "none";

  // Add the link to your DOM
  document.body.appendChild(downloadLink);

  // Lanzamos
  downloadLink.click();
}

function export_table_to_csv(html, filename) {
  var csv = [];
  let tRowsLength = tableDiv.querySelectorAll("tr").length;
  var rows = document.querySelectorAll("table tr");
  if (tRowsLength > 0) {
    for (var i = 0; i < rows.length; i++) {
      var row = [],
        cols = rows[i].querySelectorAll("td, th");

      for (var j = 0; j < cols.length; j++)
        //@ts-ignore
        row.push(cols[j].innerText);

      csv.push(row.join(","));
    }

    // Download CSV
    download_csv(csv.join("\n"), filename);
  } else {
    alert("Nothing to Export");
  }
}

// Search Logic

function search() {
  //  Spreading problems

  let problems = [...easyProblems, ...mediumProblems, ...hardProblems];
  let query = searchQueryDiv.value;
  query = query.toLowerCase();

  let searches = problems.filter((item) => {
    return item.problemName.toLowerCase().includes(query);
  });
  clearUI();
  if (searches.length > 0) {
    renderItem(searches);
  } else {
    const spanElem = document.createElement("span");
    spanElem.className = "no-items";
    spanElem.innerText = "Oops ! Nothing found";
    tableDiv.appendChild(spanElem);
  }
  searchQueryDiv.value = "";
}
